pkg/job/discovery.go

```diff
@@ -94,41 +94,63 @@ func scrapeDiscoveryJobUsingMetricData(
 	length := getMetricDataInputLength(job.Metrics)
 	partition := int(math.Ceil(float64(metricDataLength) / float64(maxMetricCount)))
 
-	mux := &sync.Mutex{}
 	var wg sync.WaitGroup
 	wg.Add(partition)
 
+	getMetricDataOutput := make([]*cloudwatch.GetMetricDataOutput, partition)
+	count := 0
+
 	for i := 0; i < metricDataLength; i += maxMetricCount {
-		go func(i int) {
+		go func(i, n int) {
 			defer wg.Done()
 			end := i + maxMetricCount
 			if end > metricDataLength {
 				end = metricDataLength
 			}
 			input := getMetricDatas[i:end]
 			filter := apicloudwatch.CreateGetMetricDataInput(input, &svc.Namespace, length, job.Delay, roundingPeriod, logger)
-			data := clientCloudwatch.GetMetricData(ctx, filter)
-			if data != nil {
-				output := make([]*model.CloudwatchData, 0, len(data.MetricDataResults))
-				for _, MetricDataResult := range data.MetricDataResults {
-					getMetricData, err := findGetMetricDataByID(input, *MetricDataResult.Id)
-					if err == nil {
-						if len(MetricDataResult.Values) != 0 {
-							getMetricData.GetMetricDataPoint = MetricDataResult.Values[0]
-							getMetricData.GetMetricDataTimestamps = MetricDataResult.Timestamps[0]
-						}
-						output = append(output, getMetricData)
-					}
-				}
-				mux.Lock()
-				cw = append(cw, output...)
-				mux.Unlock()
+			if data := clientCloudwatch.GetMetricData(ctx, filter); data != nil {
+				getMetricDataOutput[n] = data
 			}
-		}(i)
+		}(i, count)
+		count++
 	}
-
 	wg.Wait()
-	return resources, cw
+
+	// update getMetricDatas with fetched values and timestamps
+	for _, data := range getMetricDataOutput {
+		for _, metricDataResult := range data.MetricDataResults {
+			idx := findGetMetricDataByID(getMetricDatas, *metricDataResult.Id)
+			if idx == -1 {
+				logger.Warn("GetMetricData returned unknown metric ID", "metric_id", *metricDataResult.Id)
+				continue
+			}
+			if len(metricDataResult.Values) != 0 {
+				getMetricDatas[idx].GetMetricDataPoint = metricDataResult.Values[0]
+				getMetricDatas[idx].GetMetricDataTimestamps = metricDataResult.Timestamps[0]
+			}
+			getMetricDatas[idx].MetricID = nil // mark as processed
+		}
+	}
+
+	// remove unprocessed/unknown elements in place (if any)
+	getMetricDatas = compactSlice(getMetricDatas)
```

Comment:
the compaction is worth because `getMetricDatas` might have a really high cardinality for example in EC2 with lots of instances + we release early memory right?